internal/processors/enrichevents/enrichevents.go

```diff
@@ -0,0 +1,115 @@
+package enrichevents
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/DIMO-Network/cloudevent"
+	"github.com/DIMO-Network/dis/internal/processors"
+	"github.com/DIMO-Network/dis/internal/processors/cloudeventconvert"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/redpanda-data/benthos/v4/public/service"
+)
+
+type processor struct {
+	logger            *service.Logger
+	vehicleNFTAddress common.Address
+	chainID           uint64
+}
+
+// Close to fulfill the service.Processor interface.
+func (*processor) Close(context.Context) error {
+	return nil
+}
+
+// ProcessBatch to fulfill the service.BatchProcessor interface.
+func (v *processor) ProcessBatch(ctx context.Context, msgs service.MessageBatch) ([]service.MessageBatch, error) {
+	var retBatches []service.MessageBatch
+	for _, msg := range msgs {
+		retBatches = append(retBatches, v.processMsg(ctx, msg))
+	}
+	return retBatches, nil
+}
+
+func (v *processor) processMsg(_ context.Context, msg *service.Message) service.MessageBatch {
+	batch := service.MessageBatch{msg}
+	event, err := processors.MsgToEvent(msg)
+	if err != nil {
+		processors.SetError(msg, processorName, "failed to convert to event", err)
+		return batch
+	}
+
+	if event.Type != cloudevent.TypeEvent {
+		return batch
+	}
+
+	if err := v.validateEvent(event); err != nil {
+		processors.SetError(msg, processorName, "failed to validate event", err)
+		return batch
+	}
+
+	var evts Events
+	if err := json.Unmarshal(event.Data, &evts); err != nil {
+		processors.SetError(msg, processorName, "failed to parse event specific data", err)
+		return batch
+	}
+
+	if event.Extras == nil {
+		event.Extras = make(map[string]any)
+	}
+
+	for _, evt := range evts.Events {
+		if evt.Name == "" || !cloudeventconvert.ValidCharacters.MatchString(evt.Name) {
+			processors.SetError(msg, processorName, "invalid event category", fmt.Errorf("missing or invalid event category: %s", evt.Name))
+			batch = append(batch, msg)
+			continue
+		}
+
+		event.Extras[evt.Name] = &EventExtras{
+			Duration: evt.Duration,
+			Time:     evt.Time,
+		}
+	}
+
+	return batch
+}
+
+type Events struct {
+	Events []EventData `json:"events"`
+}
+
+type EventData struct {
+	Name     string          `json:"name"`
+	Time     string          `json:"time,omitempty"`
```

Comment:
Same question here actually.

Reply:
probably don't need to also validate here since we're setting these based on the above